### ISSUE TITLE
[question] What is the recommended way to enforce cppstd for requirements packages?

### ISSUE THREAD

**Issue by hrz6976** [conan-io/conan#15404]
### What is your question?

Hi there, I want to enforce the `cppstd` setting for the project and all project dependencies to `17`, as many of the dependencies do not have `gnu17` pre-built wheels available. Below is what I have tried but had no luck with:

1. Command line flags (working but have to educate others)
Enforcing others to edit their default profile or use the correct combination of command line arguments is a bit troublesome in the real world because all of us make mistakes 😅 . 

2. Profiles (Have to specify arch & compiler version)
An alternative way is to use profiles but from my experiments, arch and compiler_version are required fields. It may be intentional and can be good for reproducible builds, but the problem is we are working on different versions of gcc (varies from 9-12) on both x86 and arm machines,  it would only adds complexity to the problem if we have to maintain a long list of profiles or introduce a text preprocessor.

TLDR, I want to **set compiler and compiler.cpp_std for my project dependencies in a configuration file e.g. conanfile.py, with minimal education effort to my colleages, and let conan detect the rest.** 

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @hrz6976 

Thanks for your question.

Is it not very clear what is the issue. Conan has the ``compiler.cppstd=17`` that can be defined in profiles or in command line exactly for this goal. The ``compiler.cppstd`` will be in most cases almost compulsory in Conan 2.0, as it will be checked by recipes ``validate()`` method and raise if not defined. Why not setting ``compiler.cppstd=17`` in your profile doesn't solve the issue? It is supported by all gcc > 8, so it should be good in your default profile?

**Comment by memsharded** [comment]
Hi @hrz6976 

> it would only adds complexity to the problem if we have to maintain a long list of profiles or introduce a text preprocessor.

Any feedback here? As I suggested above, all your compiler version support cppstd=17, it shouldn't be an issue?
Even if not, using the ``detect_api`` in profiles and jinja template, you can probably encode that in just one profile if you want.

**Comment by memsharded** [comment]
I am closing this question as staled. Please re-open or create a new one if you have any further question, thanks!

**Comment by mmomtchev** [comment]
Take OpenEXR as an example - the latest version cannot be compiled unless compiling in C++17 mode - which is not the default mode on MSVC. I think that the C++ standard is a property of the source code and not the build profile. Although, in theory, you can write code that compiles in both C++17 and C++11 mode, most people do not bother. Most of the time the C++ standard of a project is set in stone and you simply have to use the right one. Besides, not counting the C++11 `std::string` epic, normally code compiled as one C++ standard should link with code compiled as another one.

Ideally, this should be something each dependency sets for itself. Most of the time, this is not a decision of the system integrator - but rather of the upstream software developer.
